really return on context done

akademic opened the pull request:
Hi, Axel

I noticed that the comment for the function doesn't match its operation. 
So, I fixed the function to work as described in the comment.

Merovius commented:
Alternatively, if that doesn't work for some reason, another possibility that might work is spawning a goroutine that waits for the context to be done and then either calls `l.SetDeadline(time.Now())` or just closes it.

I think ideally there should be a test case for this, just to reproduce it and make sure the fix actually fixes it.

akademic committed:
really return on context done

akademic commented:
> seems like a hack

I agree. My first implementation was dirty - did it in the night to get fast result in my task.
I tested approach with [(*net.ListenConfig).Listen](https://pkg.go.dev/net#ListenConfig.Listen)
It does not work as may expect - it does not finish l.Accept waiting on context.Done.

So I made new version inspired by this answer https://stackoverflow.com/a/66755998

Check if it suitable for you, or let me know how can I improve this PR

akademic committed:
added tests for cancel ListenAndServe

Merovius commented:
Yes, it seems this is golang/go#10527. Hm.

I don't think this is quite correct. If you call this with a context that never gets cancelled and then the `Accept` call fails for a *different* reason (e.g. the network goes down), this will leak a goroutine. Also, in general, a `select` with a `default` clause is a red flag. In this case, you don't need the `quit` channel, as you can use `errors.Is(err, net.ErrClosed)` to distinguish the case where the listener was closed.

I think the steps needed are

1. Use `ctx, cancel := context.WithCancel(ctx); defer cancel()` to get a child context that definitely gets cancelled eventually. Do that after the `defer wg.Done()`, so the cancellation happens *before* we wait on the `WaitGroup`. i.e. move the `WaitGroup` stuff to the very top of the function.
2. Put the goroutine that does the closing into the same `WaitGroup`. We want to guarantee that the listener is closed before returning. This is also why it's important to `defer wg.Done()` before the `defer cancel()`.

akademic committed:
fixed goroutine leak when exit on error

akademic commented:
Here are fixes you requested

akademic committed:
enshure that connection is closed before returning from ListenAndServe

Merovius commented on transmission.go:
```diff
@@ -48,15 +48,30 @@ type Device interface {
+	go func() {
+		<-ctxCancel.Done()
+		l.Close()
+		wg.Done()
```
Please `defer wg.Done()` at the top of the function. It's unlikely to make a semantic difference, but it's better to stay with the consistent pattern.

akademic replied:
ok

Merovius commented on transmission_test.go:
```diff
@@ -0,0 +1,70 @@
+	// Start the server
+	exited := make(chan any)
+	go func() {
+		lErr := ListenAndServe(ctx, "unix", sockFile, Export{})
```
Just call this `err`.

akademic replied:
ok

Merovius commented on transmission_test.go:
```diff
@@ -0,0 +1,70 @@
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	dir, _ := os.MkdirTemp("", "")
```
`dir := t.TempDir()`

akademic replied:
Didn't know about this. Thanx a lot, it will make my tests better.

Merovius commented on transmission_test.go:
```diff
@@ -0,0 +1,70 @@
+	// Start the server
+	exited := make(chan any)
+	go func() {
+		lErr := ListenAndServe(ctx, "unix", sockFile, Export{})
```
Just `err`

akademic replied:
ok

Merovius commented on transmission_test.go:
```diff
@@ -0,0 +1,70 @@
+		if lErr != nil {
+			t.Errorf("ListenAndServe returned an error: %v", lErr)
+		}
+		close(exited)
```
Make this `defer close(exited)` at the top of the function (guard against `ListenAndServe` panicing in a future change).

akademic replied:
Good point

Merovius commented on transmission_test.go:
```diff
@@ -0,0 +1,70 @@
+		if lErr != nil {
+			t.Errorf("ListenAndServe returned an error: %v", lErr)
+		}
+		close(exited)
```
`defer close(exited)` at the top here as well.

akademic replied:
ok

Merovius commented on transmission_test.go:
```diff
@@ -0,0 +1,70 @@
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	dir, _ := os.MkdirTemp("", "")
```
`dir := t.TempDir()`

akademic replied:
ok

Merovius reviewed (commented):
Apologies for the delays. 

Overall this looks good, but there are a couple of nits. I realize that it is a bit strange to become nitpicky, given that the rest of the code is probably not entirely tidy (I assume, given how old it is). But better late, than never. I really appreciate your contribution.

akademic committed:
consistent use of wg.Done()

akademic committed:
make tests better

- make temp dir with testing.T.TempDir
- make tests working in case of panic
- better naming

akademic commented:
Here commits with changes after review.
As for code quality, my philosophy is that if project works better after patch, then everything goes ok

Merovius commented:
Thanks again.
